### ISSUE TITLE
[Client Side] implement the practitioner details endpoint. 

### ISSUE THREAD

**Issue by dubdabasoduba** [opensrp/fhircore#857]
**Is your feature request related to a problem? Please describe.**
- After creating a [practitioner details fetch endpoint](https://github.com/opensrp/fhircore/issues/853) we need to get, save & use the data in FHIR core. 

**Describe the solution you'd like**
**Getting data**
- [x] Implement a GET of the [practitioner details fetch endpoint](https://github.com/opensrp/fhircore/issues/853)  as the 1st endpoint call after a user performs a remote login.  
     -  The endpoint parameter should be the user/practitioner id from keycloak.  i.e the keycloak uuid

**Saving data**
- [x] Currently, we have some keycloak response saved in the secure shared preferences. Update this to save all the keycloak info included in the [practitioner details fetch endpoint](https://github.com/opensrp/fhircore/issues/853) response. 
- [x] Save all the resources in the payload mentioned above in the `ResourceEntity` table. 

**Using data**
- [x] Write helper methods to fetch the keycloak related data from the secure shared preferences
      - Should be able to get all demographics
      - Should be able to get all the roles assigned to the user. 
- [x] Write helper a method to fetch all the practitioner-related resources.
      - This can be generic i.e give it the `practitioner-id `and `resource type` then it returns the exact resource required. 

**Additional context**
- [Sample payload](https://github.com/opensrp/fhircore/issues/853)

**Comment by pld** [comment]
Is there anything needed here that we can architect to be reusable in a supervisor app that fetches the details of a set of practitioners?

**Comment by RaaziaTarique** [comment]
We tried the possibility to add a custom UserDetails Resource class
```
import org.hl7.fhir.r4.model.Resource;
import org.hl7.fhir.r4.model.ResourceType;

public class UserDetails extends Resource {
  @Override
  public ResourceType getResourceType() {
    return ResourceType.Basic;
  }
}
```

But you see we need to extend getResourceType() method and it can only use ResourceType which is only available in the Fhir library. Even if we want make modifications into extending resource type we need to pre-defined ResourceType.

<img width="1092" alt="Screenshot 2021-12-24 at 1 09 42 PM" src="https://user-images.githubusercontent.com/89384079/147351642-cbe8246e-2ca3-4f79-9268-76c06bcefc04.png">

@dubdabasoduba @rehammuzzamil @maimoonak

**Comment by rehammuzzamil** [comment]
> We tried the possibility to add a custom UserDetails Resource class
> 
> ```
> import org.hl7.fhir.r4.model.Resource;
> import org.hl7.fhir.r4.model.ResourceType;
> 
> public class UserDetails extends Resource {
>   @Override
>   public ResourceType getResourceType() {
>     return ResourceType.Basic;
>   }
> }
> ```
> 
> But you see we need to extend getResourceType() method and it can only use ResourceType which is only available in the Fhir library. Even if we want make modifications into extending resource type we need to use pre-defined ResourceType.
> 
> <img alt="Screenshot 2021-12-24 at 1 09 42 PM" width="1092" src="https://user-images.githubusercontent.com/89384079/147351642-cbe8246e-2ca3-4f79-9268-76c06bcefc04.png">
> 
> @dubdabasoduba @rehammuzzamil @maimoonak

@RaaziaTarique As discussed, we need to register our custom Resource Provider in the FHIR. On the server-side, we achieve this by creating a new class with the `@ResourceDef `annotation. Once the class is created, we need to register our new Resource Provider (custom class) and other supporting classes in the Restful server. Registration of Resource Provider can be done by using `registerProvider(practitionerDetailsResourceProvider); `method at the backend.

Following is the example of custom resource provider:
```

@ResourceDef(
        name = "practitioner-details",
        profile = "http://hl7.org/fhir/profiles/custom-resource")
public class PractitionerDetails extends Practitioner {

    @Child(
            name = "KeycloakUserDetails",
            type = {KeycloakUserDetails.class})
    @Description(
            shortDefinition = "Complete KeycloakUserDetails",
            formalDefinition = "KeycloakUserDetails")
    private KeycloakUserDetails keycloakUserDetails;

    @Child(
            name = "fhir",
            type = {FhirPractitionerDetails.class})
    @Description(
            shortDefinition = "Get data from FHIR Server",
            formalDefinition = "Get data from FHIR Server")
    private FhirPractitionerDetails fhirPractitionerDetails;

    @Override
    public Practitioner copy() {
        Practitioner practitioner = new Practitioner();
        Bundle bundle = new Bundle();
        List<Bundle.BundleEntryComponent> theEntry = new ArrayList<>();
        Bundle.BundleEntryComponent entryComponent = new Bundle.BundleEntryComponent();
        entryComponent.setResource(new Bundle());
        theEntry.add(entryComponent);
        bundle.setEntry(theEntry);
        this.copyValues(practitioner);
        return practitioner;
    }

    @Override
    public ResourceType getResourceType() {
        return ResourceType.Bundle;
    }
  //getters and setters for the child fields
```
}

Since the work is in progress. It is likely to be changed. 
cc: @dubdabasoduba  @maimoonak

**Comment by rehammuzzamil** [comment]
@RaaziaTarique  The error may also occur due to the server-side work not being deployed yet. It is throwing a `DataFormatException` because the FHIR Server does not provide support of the end-point yet, hence it's not understandable by FHIR. 
cc : @maimoonak  @dubdabasoduba

**Comment by dubdabasoduba** [comment]
> @RaaziaTarique The error may also occur due to the server-side work not being deployed yet. It is throwing a `DataFormatException` because the FHIR Server does not provide support of the end-point yet, hence it's not understandable by FHIR. cc : @maimoonak @dubdabasoduba

This makes sense. @rehammuzzamil when can we have bits of this done?

**Comment by rehammuzzamil** [comment]
> > @RaaziaTarique The error may also occur due to the server-side work not being deployed yet. It is throwing a `DataFormatException` because the FHIR Server does not provide support of the end-point yet, hence it's not understandable by FHIR. cc : @maimoonak @dubdabasoduba
> 
> This makes sense. @rehammuzzamil when can we have bits of this done?

@dubdabasoduba we can get an initial API PR ready by EOD Friday.

**Comment by rehammuzzamil** [comment]
The server-side PR is ready for review https://github.com/opensrp/hapi-fhir-opensrp-extensions/pull/12 
(Unit tests will be added)
cc : @dubdabasoduba  @f-odhiambo  @maimoonak

**Comment by RaaziaTarique** [comment]
> The server-side PR is ready for review [opensrp/hapi-fhir-opensrp-extensions#12](https://github.com/opensrp/hapi-fhir-opensrp-extensions/pull/12) (Unit tests will be added) cc : @dubdabasoduba @f-odhiambo @maimoonak

I have used this PR's Resource classes and I am still getting same error as previous

```
E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: org.smartregister.fhircore.anc, PID: 21132
    ca.uhn.fhir.parser.DataFormatException: Unknown resource name "practitioner-details" (this name is not known in FHIR version "R4")
        at ca.uhn.fhir.context.FhirContext.getResourceDefinition(FhirContext.java:547)
        at ca.uhn.fhir.context.FhirContext.getResourceDefinition(FhirContext.java:463)
        at ca.uhn.fhir.parser.ParserState$PreResourceState.enteringNewElement(ParserState.java:994)
        at ca.uhn.fhir.parser.ParserState.enteringNewElement(ParserState.java:136)
        at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1076)
        at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:996)
        at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1079)
        at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:1060)
        at ca.uhn.fhir.parser.JsonParser.parseChildren(JsonParser.java:996)
        at ca.uhn.fhir.parser.JsonParser.doParseResource(JsonParser.java:214)
        at ca.uhn.fhir.parser.JsonParser.doParseResource(JsonParser.java:196)
        at ca.uhn.fhir.parser.BaseParser.parseResource(BaseParser.java:603)
        at ca.uhn.fhir.parser.BaseParser.parseResource(BaseParser.java:651)
        at ca.uhn.fhir.parser.BaseParser.parseResource(BaseParser.java:661)
        at org.smartregister.fhircore.engine.auth.AccountAuthenticator.getPractitionerDetails(AccountAuthenticator.kt:189)
        at org.smartregister.fhircore.engine.ui.login.LoginViewModel$callPractitionerDetails$1.invokeSuspend(LoginViewModel.kt:113)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

My code for registering custom classes (currently a sample payload is being used as the server side PR is still not merged)

```
suspend fun getPractitionerDetails(keycloak_uuid: String): org.hl7.fhir.r4.model.Bundle {
  val iParser: IParser = FhirContext.forR4().newJsonParser()
  FhirContext.forR4().registerCustomType(PractitionerDetails::class.java)
  FhirContext.forR4().registerCustomType(FhirCareTeamExtension::class.java)
  FhirContext.forR4().registerCustomType(FhirOrganizationExtension::class.java)
  FhirContext.forR4().registerCustomType(FhirPractitionerDetails::class.java)
  FhirContext.forR4().registerCustomType(KeycloakUserDetails::class.java)
  FhirContext.forR4().registerCustomType(UserBioData::class.java)
  val qJson =
    context.assets.open("sample_practitionar_payload.json").bufferedReader().use { it.readText() }

  return iParser.parseResource(qJson) as org.hl7.fhir.r4.model.Bundle
}
```

CC: @rehammuzzamil @maimoonak  @dubdabasoduba

**Comment by dubdabasoduba** [comment]
@RaaziaTarique Please use this Parameter resources sample to hold the Practitioner resource linkage. 

```
{
  "resourceType": "Parameters",
  "parameter": [
    {
      "name": "practitioner",
      "valueString": "003a78c1-a589-435d-8d28-12b3a77b471c", --- Practitioner identifier
      "part": [
        {
          "name": "Organization",
          "valueString": "214" 
        },
        {
          "name": "CareTeam",
          "valueString": "21"
        },
        {
          "name": "CareTeam",
          "valueString": "213"
        }
      ]
    }
  ]
}

```

- In the `part` attribute we have an array of objects. The attributes in each object are to  be used as follows
    - `name` -- this will hold the resource type
    - `valueString` -- This will hold the resource ID. 

- After the generation of this Parameter resource we will save it and get its ID. 
- We will then save this ID in the secure shared preferences under the tag `practitioner_details`

**Comment by RaaziaTarique** [comment]
@f-odhiambo Library for Practitioner details API models is deployed but the API is currently working locally only. @rehammuzzamil needs to check logs inorder to investigate the reason of failure of the API on FHIR. Meanwhile I have removed the models that I copied in the project and used the ones that are deployed here https://oss.sonatype.org/content/repositories/snapshots/org/smartregister/fhir-common-utils/0.0.1-SNAPSHOT/). I will continue working on this after finalising issue https://github.com/opensrp/fhircore/issues/981

**Comment by RaaziaTarique** [comment]
While saving the custom resources `FhirCareTeamExtension`, `FhirOrganizationExtension` and `LocationHierarchy` into the `ResourceEntity` table, I am getting the following error `java.lang.ClassCastException: org.hl7.fhir.r4.model.CareTeam cannot be cast to org.smartregister.model.practitioner.FhirCareTeamExtension` 
`java.lang.ClassCastException: class org.hl7.fhir.r4.model.Organization cannot be cast to class org.smartregister.model.practitioner.FhirOrganizationExtension` because in `PractitionerDetails` we are using `FhirCareTeamExtension`, `FhirOrganizationExtension` and  the server side response is returning list of  `Organization ` and `CareTeam `. 
If I try to modify the the saved json response to return list of `FhirCareTeamExtension`, `FhirOrganizationExtension` then it throws an exception saying that `FhirCareTeamExtension`, `FhirOrganizationExtension` and `LocationHierarchy` are not recognised by `FHIR version R4`. 
Right now either can change the `PractitionerDetails` class to use FHIR known resources or to modify FHIR SDK but for now I think it's better to use FHIR known resources .
@maimoonak @dubdabasoduba @rehammuzzamil CC: @f-odhiambo

**Comment by rehammuzzamil** [comment]
@RaaziaTarique  Is there a client-side way to store custom resources in the 'ResourceEntity' table?
We used custom resources because we couldn't use FHIR Resources directly in a custom response payload/entity if we wanted to include them as a child. The error "@child annotation cannot be applied to IBaseResource" was thrown.

cc : @dubdabasoduba

**Comment by RaaziaTarique** [comment]
> @RaaziaTarique Is there a client-side way to store custom resources in the 'ResourceEntity' table? We used custom resources because we couldn't use FHIR Resources directly in a custom response payload/entity if we wanted to include them as a child. The error "@child annotation cannot be applied to IBaseResource" was thrown.
> 
> cc : @dubdabasoduba

In order to do that we need to make FHIR SDK understand custom resources other then `FHIR version R4` resources.  @rehammuzzamil @maimoonak @dubdabasoduba

**Comment by dubdabasoduba** [comment]
> > @RaaziaTarique Is there a client-side way to store custom resources in the 'ResourceEntity' table? We used custom resources because we couldn't use FHIR Resources directly in a custom response payload/entity if we wanted to include them as a child. The error "@child annotation cannot be applied to IBaseResource" was thrown.
> > cc : @dubdabasoduba
> 
> In order to do that we need to make FHIR SDK understand custom resources other then `FHIR version R4` resources. @rehammuzzamil @maimoonak @dubdabasoduba

@RaaziaTarique I don't think this is something we can add on SDK. We might need to extend it and write the custom pieces on FHIR core instead.

**Comment by RaaziaTarique** [comment]
@dubdabasoduba if somehow we'll extend `FhirEngine` to write custom resources then there will still be one error remaining which is type casting of `CareTeam` and `Organization` to `FhirCareTeamExtension` and `FhirOrganizationExtension` respectively. @rehammuzzamil @maimoonak

**Comment by rehammuzzamil** [comment]
> While saving the custom resources `FhirCareTeamExtension`, `FhirOrganizationExtension` and `LocationHierarchy` into the `ResourceEntity` table, I am getting the following error `java.lang.ClassCastException: org.hl7.fhir.r4.model.CareTeam cannot be cast to org.smartregister.model.practitioner.FhirCareTeamExtension` `java.lang.ClassCastException: class org.hl7.fhir.r4.model.Organization cannot be cast to class org.smartregister.model.practitioner.FhirOrganizationExtension` because in `PractitionerDetails` we are using `FhirCareTeamExtension`, `FhirOrganizationExtension` and the server side response is returning list of `Organization ` and `CareTeam `. If I try to modify the the saved json response to return list of `FhirCareTeamExtension`, `FhirOrganizationExtension` then it throws an exception saying that `FhirCareTeamExtension`, `FhirOrganizationExtension` and `LocationHierarchy` are not recognised by `FHIR version R4`. Right now either can change the `PractitionerDetails` class to use FHIR known resources or to modify FHIR SDK but for now I think it's better to use FHIR known resources . @maimoonak @dubdabasoduba @rehammuzzamil CC: @f-odhiambo

@RaaziaTarique , Server side response is already returning a list of FhirCareTeamExtension, a list of FhirOrganizationExtension and a list of LocationHierarchy in a FhirPractitionerDetails custom class. You can refer to it [here](https://github.com/opensrp/fhir-common-utils/blob/7c9106ce5db6fbb84533f2efd34d3227c6c9461f/src/main/java/org/smartregister/model/practitioner/FhirPractitionerDetails.java#L29). 

cc : @dubdabasoduba

**Comment by rehammuzzamil** [comment]
Ideally, if we provide support to persist custom classes then it should not throw class cast exceptions. As the server returns the response payload in the custom classes structure. 

Please correct me if I am missing something. Thanks!
@RaaziaTarique 

cc : @maimoonak  @dubdabasoduba

**Comment by rehammuzzamil** [comment]
@RaaziaTarique  As per discussion, we have updated the model classes to use FHIR known classes instead of custom classes for Care Teams and Organizations. We will also be returning a list of locations with which the practitioner is associated in the response payload.  
Here is the PR https://github.com/opensrp/fhir-common-utils/pull/4 for reference.

cc : @dubdabasoduba  @maimoonak  @f-odhiambo

**Comment by f-odhiambo** [comment]
I had a catch up with @dubdabasoduba on this and also @RaaziaTarique . This looks good. Finally taking shape 
Also, thanks for the updates and also the comprehensive work on this @rehammuzzamil @RaaziaTarique @dubdabasoduba Looking forward to incorporate this to clinical workflows . I think a demo would also be nice even if its recorded .

I see also https://github.com/opensrp/fhir-common-utils/pull/4 has been merged

**Comment by rehammuzzamil** [comment]
Thanks @f-odhiambo .
The PR https://github.com/opensrp/fhir-common-utils/pull/4 is merged, now we need the tag to be published. 
https://github.com/opensrp/fhir-common-utils/releases/tag/v0.0.2-SNAPSHOT   

Sure, we can plan a demo for this.
cc : @dubdabasoduba

**Comment by f-odhiambo** [comment]
Part of this task has been refactored and partly assigned to @ellykits

**Comment by f-odhiambo** [comment]
@RaaziaTarique Kindly share the official payload for this CC @dubdabasoduba , Also should @FikriMilano build on this PR https://github.com/opensrp/fhircore/pull/890 ?

**Comment by rehammuzzamil** [comment]
Here is the sample response payload:
`{
  "resourceType": "Bundle",
  "id": "1c3e2cfe-cea0-470e-a032-1ccf39431fd8",
  "meta": {
    "lastUpdated": "2022-08-06T12:48:22.920+00:00"
  },
  "type": "searchset",
  "total": 1,
  "link": [ {
    "relation": "self",
    "url": ["https://fhir.labs.smartregister.org:443/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9"](https://fhir.labs.smartregister.org/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9)
  } ],
  "entry": [ {
    "fullUrl": ["https://fhir.labs.smartregister.org:443/fhir/practitioner-details/34f0d616-afc8-4446-ae87-bb60be4bdbc9"](https://fhir.labs.smartregister.org/fhir/practitioner-details/34f0d616-afc8-4446-ae87-bb60be4bdbc9),
    "resource": {
      "resourceType": "practitioner-details",
      "id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
      "meta": {
        "profile": [ "http://hl7.org/fhir/profiles/custom-resource" ]
      },
      "KeycloakUserDetails": {
        "id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
        "user-bio": [ {
          "identifier": "b87ff3c2-cbc6-43e6-b753-a9620756f9e4",
          "userName": "demo",
          "preferredName": "demo",
          "familyName": "16909",
          "givenName": "Demo",
          "emailVerified": "false"
        } ],
        "user-roles": [ "ROLE_realm-admin", "ROLE_OPENMRS", "ROLE_EDIT_KEYCLOAK_USERS", "ROLE_offline_access", "ROLE_VIEW_KEYCLOAK_USERS", "ROLE_uma_authorization", "ROLE_ALL_EVENTS", "ROLE_PLANS_FOR_USER" ]
      },
      "fhir": {
        "id": "136252",
        "careteams": [ {
          "resourceType": "CareTeam",
          "id": "136253",
          "meta": {
            "versionId": "1",
            "lastUpdated": "2022-07-25T16:41:45.221+00:00",
            "source": "#98d1c5ca256d7e8b"
          },
          "identifier": [ {
            "use": "official",
            "value": "46384571-ad31-4c20-8aa0-266f80ef6582"
          } ],
          "status": "active",
          "name": "Nala Team",
          "participant": [ {
            "member": {
              "reference": ["Practitioner/136252"](https://fhir.labs.smartregister.org/fhir/Practitioner/136252)
            }
          } ]
        } ],
        "teams": [ {
          "resourceType": "Organization",
          "id": "136254",
          "meta": {
            "versionId": "2",
            "lastUpdated": "2022-07-25T16:42:51.523+00:00",
            "source": "#dad854aca8ea9fcd"
          },
          "identifier": [ {
            "use": "official",
            "value": "e58c9509-8ff2-4664-b805-e9dd5bf0cf8a"
          } ],
          "active": true,
          "type": [ {
            "coding": [ {
              "system": ["http://terminology.hl7.org/CodeSystem/organization-type"](http://terminology.hl7.org/CodeSystem/organization-type),
              "code": "prov"
            } ]
          } ],
          "name": "Nala Team",
          "alias": [ "nala" ]
        } ],
        "locations": [ {
          "resourceType": "Location",
          "id": "136256",
          "meta": {
            "versionId": "1",
            "lastUpdated": "2022-07-25T16:44:38.620+00:00",
            "source": "#b46640c1adfc6d38"
          },
          "identifier": [ {
            "use": "official",
            "value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
          } ],
          "status": "active",
          "name": "Nala Location ",
          "physicalType": {
            "coding": [ {
              "system": ["http://terminology.hl7.org/CodeSystem/location-physical-type"](http://terminology.hl7.org/CodeSystem/location-physical-type),
              "code": "jdn",
              "display": "Jurisdiction"
            } ]
          },
          "partOf": {
            "reference": ["Location/109211"](https://fhir.labs.smartregister.org/fhir/Location/109211),
            "display": "Charlie Clinic"
          }
        } ],
        "locationHierarchyList": [ {
          "resourceType": "LocationHierarchy",
          "id": "Location Resource : 136256",
          "meta": {
            "profile": [ "http://hl7.org/fhir/profiles/custom-resource" ]
          },
          "LocationHierarchyTree": {
            "locationsHierarchy": {
              "listOfNodes": {
                "treeNodeId": ["Location/136256"](https://fhir.labs.smartregister.org/fhir/Location/136256),
                "treeNode": [ {
                  "nodeId": ["Location/136256"](https://fhir.labs.smartregister.org/fhir/Location/136256),
                  "label": "Nala Location ",
                  "node": {
                    "resourceType": "Location",
                    "id": "136256",
                    "meta": {
                      "versionId": "1",
                      "lastUpdated": "2022-07-25T16:44:38.620+00:00",
                      "source": "#b46640c1adfc6d38"
                    },
                    "identifier": [ {
                      "use": "official",
                      "value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
                    } ],
                    "status": "active",
                    "name": "Nala Location ",
                    "physicalType": {
                      "coding": [ {
                        "system": ["http://terminology.hl7.org/CodeSystem/location-physical-type"](http://terminology.hl7.org/CodeSystem/location-physical-type),
                        "code": "jdn",
                        "display": "Jurisdiction"
                      } ]
                    },
                    "partOf": {
                      "reference": ["Location/109211"](https://fhir.labs.smartregister.org/fhir/Location/109211),
                      "display": "Charlie Clinic"
                    }
                  },
                  "parent": ["Location/109211"](https://fhir.labs.smartregister.org/fhir/Location/109211),
                  "children": [ {
                    "childId": ["Location/137151"](https://fhir.labs.smartregister.org/fhir/Location/137151),
                    "treeNode": {
                      "nodeId": ["Location/137151"](https://fhir.labs.smartregister.org/fhir/Location/137151),
                      "label": "Test Village",
                      "node": {
                        "resourceType": "Location",
                        "id": "137151",
                        "meta": {
                          "versionId": "1",
                          "lastUpdated": "2022-08-04T11:41:32.683+00:00",
                          "source": "#1bb63352eef7cefc"
                        },
                        "identifier": [ {
                          "use": "official",
                          "value": "f7f4c729-fd05-40aa-bdc3-fccd74589264"
                        } ],
                        "status": "active",
                        "name": "Test Village",
                        "description": "Test",
                        "physicalType": {
                          "coding": [ {
                            "system": ["http://terminology.hl7.org/CodeSystem/location-physical-type"](http://terminology.hl7.org/CodeSystem/location-physical-type),
                            "code": "jdn",
                            "display": "Jurisdiction"
                          } ]
                        },
                        "partOf": {
                          "reference": ["Location/136256"](https://fhir.labs.smartregister.org/fhir/Location/136256),
                          "display": "Nala Location "
                        }
                      },
                      "parent": ["Location/136256"](https://fhir.labs.smartregister.org/fhir/Location/136256)
                    }
                  } ]
                } ]
              },
              "parentChildren": [ {
                "identifier": ["Location/109211"](https://fhir.labs.smartregister.org/fhir/Location/109211),
                "childIdentifiers": [ "Location/136256" ]
              }, {
                "identifier": ["Location/136256"](https://fhir.labs.smartregister.org/fhir/Location/136256),
                "childIdentifiers": [ "Location/137151" ]
              } ]
            }
          },
          "locationId": "136256"
        } ],
        "practitionerId": [ "136252" ]
      }
    }
  } ]
}`

You can use an endpoint https://fhir.labs.smartregister.org/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9 for testing purposes. @f-odhiambo 
cc : @dubdabasoduba

**Comment by rehammuzzamil** [comment]
`{
	"resourceType": "Bundle",
	"id": "1c3e2cfe-cea0-470e-a032-1ccf39431fd8",
	"meta": {
		"lastUpdated": "2022-08-06T12:48:22.920+00:00"
	},
	"type": "searchset",
	"total": 1,
	"link": [{
		"relation": "self",
		"url": "https://fhir.labs.smartregister.org:443/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9"
	}],
	"entry": [{
		"fullUrl": "https://fhir.labs.smartregister.org:443/fhir/practitioner-details/34f0d616-afc8-4446-ae87-bb60be4bdbc9",
		"resource": {
			"resourceType": "practitioner-details",
			"id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
			"meta": {
				"profile": ["http://hl7.org/fhir/profiles/custom-resource"]
			},
			"KeycloakUserDetails": {
				"id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
				"user-bio": [{
					"identifier": "b87ff3c2-cbc6-43e6-b753-a9620756f9e4",
					"userName": "demo",
					"preferredName": "demo",
					"familyName": "16909",
					"givenName": "Demo",
					"emailVerified": "false"
				}],
				"user-roles": ["ROLE_realm-admin", "ROLE_OPENMRS", "ROLE_EDIT_KEYCLOAK_USERS", "ROLE_offline_access", "ROLE_VIEW_KEYCLOAK_USERS", "ROLE_uma_authorization", "ROLE_ALL_EVENTS", "ROLE_PLANS_FOR_USER"]
			},
			"fhir": {
				"id": "136252",
				"careteams": [{
					"resourceType": "CareTeam",
					"id": "136253",
					"meta": {
						"versionId": "1",
						"lastUpdated": "2022-07-25T16:41:45.221+00:00",
						"source": "#98d1c5ca256d7e8b"
					},
					"identifier": [{
						"use": "official",
						"value": "46384571-ad31-4c20-8aa0-266f80ef6582"
					}],
					"status": "active",
					"name": "Nala Team",
					"participant": [{
						"member": {
							"reference": "Practitioner/136252"
						}
					}]
				}],
				"teams": [{
					"resourceType": "Organization",
					"id": "136254",
					"meta": {
						"versionId": "2",
						"lastUpdated": "2022-07-25T16:42:51.523+00:00",
						"source": "#dad854aca8ea9fcd"
					},
					"identifier": [{
						"use": "official",
						"value": "e58c9509-8ff2-4664-b805-e9dd5bf0cf8a"
					}],
					"active": true,
					"type": [{
						"coding": [{
							"system": "http://terminology.hl7.org/CodeSystem/organization-type",
							"code": "prov"
						}]
					}],
					"name": "Nala Team",
					"alias": ["nala"]
				}],
				"locations": [{
					"resourceType": "Location",
					"id": "136256",
					"meta": {
						"versionId": "1",
						"lastUpdated": "2022-07-25T16:44:38.620+00:00",
						"source": "#b46640c1adfc6d38"
					},
					"identifier": [{
						"use": "official",
						"value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
					}],
					"status": "active",
					"name": "Nala Location ",
					"physicalType": {
						"coding": [{
							"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
							"code": "jdn",
							"display": "Jurisdiction"
						}]
					},
					"partOf": {
						"reference": "Location/109211",
						"display": "Charlie Clinic"
					}
				}],
				"locationHierarchyList": [{
					"resourceType": "LocationHierarchy",
					"id": "Location Resource : 136256",
					"meta": {
						"profile": ["http://hl7.org/fhir/profiles/custom-resource"]
					},
					"LocationHierarchyTree": {
						"locationsHierarchy": {
							"listOfNodes": {
								"treeNodeId": "Location/136256",
								"treeNode": [{
									"nodeId": "Location/136256",
									"label": "Nala Location ",
									"node": {
										"resourceType": "Location",
										"id": "136256",
										"meta": {
											"versionId": "1",
											"lastUpdated": "2022-07-25T16:44:38.620+00:00",
											"source": "#b46640c1adfc6d38"
										},
										"identifier": [{
											"use": "official",
											"value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
										}],
										"status": "active",
										"name": "Nala Location ",
										"physicalType": {
											"coding": [{
												"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
												"code": "jdn",
												"display": "Jurisdiction"
											}]
										},
										"partOf": {
											"reference": "Location/109211",
											"display": "Charlie Clinic"
										}
									},
									"parent": "Location/109211",
									"children": [{
										"childId": "Location/137151",
										"treeNode": {
											"nodeId": "Location/137151",
											"label": "Test Village",
											"node": {
												"resourceType": "Location",
												"id": "137151",
												"meta": {
													"versionId": "1",
													"lastUpdated": "2022-08-04T11:41:32.683+00:00",
													"source": "#1bb63352eef7cefc"
												},
												"identifier": [{
													"use": "official",
													"value": "f7f4c729-fd05-40aa-bdc3-fccd74589264"
												}],
												"status": "active",
												"name": "Test Village",
												"description": "Test",
												"physicalType": {
													"coding": [{
														"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
														"code": "jdn",
														"display": "Jurisdiction"
													}]
												},
												"partOf": {
													"reference": "Location/136256",
													"display": "Nala Location "
												}
											},
											"parent": "Location/136256"
										}
									}]
								}]
							},
							"parentChildren": [{
								"identifier": "Location/109211",
								"childIdentifiers": ["Location/136256"]
							}, {
								"identifier": "Location/136256",
								"childIdentifiers": ["Location/137151"]
							}]
						}
					},
					"locationId": "136256"
				}],
				"practitionerId": ["136252"]
			}
		}
	}]
}`

**Comment by rehammuzzamil** [comment]
{
	"resourceType": "Bundle",
	"id": "1c3e2cfe-cea0-470e-a032-1ccf39431fd8",
	"meta": {
		"lastUpdated": "2022-08-06T12:48:22.920+00:00"
	},
	"type": "searchset",
	"total": 1,
	"link": [{
		"relation": "self",
		"url": "https://fhir.labs.smartregister.org:443/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9"
	}],
	"entry": [{
		"fullUrl": "https://fhir.labs.smartregister.org:443/fhir/practitioner-details/34f0d616-afc8-4446-ae87-bb60be4bdbc9",
		"resource": {
			"resourceType": "practitioner-details",
			"id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
			"meta": {
				"profile": ["http://hl7.org/fhir/profiles/custom-resource"]
			},
			"KeycloakUserDetails": {
				"id": "34f0d616-afc8-4446-ae87-bb60be4bdbc9",
				"user-bio": [{
					"identifier": "b87ff3c2-cbc6-43e6-b753-a9620756f9e4",
					"userName": "demo",
					"preferredName": "demo",
					"familyName": "16909",
					"givenName": "Demo",
					"emailVerified": "false"
				}],
				"user-roles": ["ROLE_realm-admin", "ROLE_OPENMRS", "ROLE_EDIT_KEYCLOAK_USERS", "ROLE_offline_access", "ROLE_VIEW_KEYCLOAK_USERS", "ROLE_uma_authorization", "ROLE_ALL_EVENTS", "ROLE_PLANS_FOR_USER"]
			},
			"fhir": {
				"id": "136252",
				"careteams": [{
					"resourceType": "CareTeam",
					"id": "136253",
					"meta": {
						"versionId": "1",
						"lastUpdated": "2022-07-25T16:41:45.221+00:00",
						"source": "#98d1c5ca256d7e8b"
					},
					"identifier": [{
						"use": "official",
						"value": "46384571-ad31-4c20-8aa0-266f80ef6582"
					}],
					"status": "active",
					"name": "Nala Team",
					"participant": [{
						"member": {
							"reference": "Practitioner/136252"
						}
					}]
				}],
				"teams": [{
					"resourceType": "Organization",
					"id": "136254",
					"meta": {
						"versionId": "2",
						"lastUpdated": "2022-07-25T16:42:51.523+00:00",
						"source": "#dad854aca8ea9fcd"
					},
					"identifier": [{
						"use": "official",
						"value": "e58c9509-8ff2-4664-b805-e9dd5bf0cf8a"
					}],
					"active": true,
					"type": [{
						"coding": [{
							"system": "http://terminology.hl7.org/CodeSystem/organization-type",
							"code": "prov"
						}]
					}],
					"name": "Nala Team",
					"alias": ["nala"]
				}],
				"locations": [{
					"resourceType": "Location",
					"id": "136256",
					"meta": {
						"versionId": "1",
						"lastUpdated": "2022-07-25T16:44:38.620+00:00",
						"source": "#b46640c1adfc6d38"
					},
					"identifier": [{
						"use": "official",
						"value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
					}],
					"status": "active",
					"name": "Nala Location ",
					"physicalType": {
						"coding": [{
							"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
							"code": "jdn",
							"display": "Jurisdiction"
						}]
					},
					"partOf": {
						"reference": "Location/109211",
						"display": "Charlie Clinic"
					}
				}],
				"locationHierarchyList": [{
					"resourceType": "LocationHierarchy",
					"id": "Location Resource : 136256",
					"meta": {
						"profile": ["http://hl7.org/fhir/profiles/custom-resource"]
					},
					"LocationHierarchyTree": {
						"locationsHierarchy": {
							"listOfNodes": {
								"treeNodeId": "Location/136256",
								"treeNode": [{
									"nodeId": "Location/136256",
									"label": "Nala Location ",
									"node": {
										"resourceType": "Location",
										"id": "136256",
										"meta": {
											"versionId": "1",
											"lastUpdated": "2022-07-25T16:44:38.620+00:00",
											"source": "#b46640c1adfc6d38"
										},
										"identifier": [{
											"use": "official",
											"value": "c3bd4bcc-889e-4e3d-a72b-a0b611b8fb64"
										}],
										"status": "active",
										"name": "Nala Location ",
										"physicalType": {
											"coding": [{
												"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
												"code": "jdn",
												"display": "Jurisdiction"
											}]
										},
										"partOf": {
											"reference": "Location/109211",
											"display": "Charlie Clinic"
										}
									},
									"parent": "Location/109211",
									"children": [{
										"childId": "Location/137151",
										"treeNode": {
											"nodeId": "Location/137151",
											"label": "Test Village",
											"node": {
												"resourceType": "Location",
												"id": "137151",
												"meta": {
													"versionId": "1",
													"lastUpdated": "2022-08-04T11:41:32.683+00:00",
													"source": "#1bb63352eef7cefc"
												},
												"identifier": [{
													"use": "official",
													"value": "f7f4c729-fd05-40aa-bdc3-fccd74589264"
												}],
												"status": "active",
												"name": "Test Village",
												"description": "Test",
												"physicalType": {
													"coding": [{
														"system": "http://terminology.hl7.org/CodeSystem/location-physical-type",
														"code": "jdn",
														"display": "Jurisdiction"
													}]
												},
												"partOf": {
													"reference": "Location/136256",
													"display": "Nala Location "
												}
											},
											"parent": "Location/136256"
										}
									}]
								}]
							},
							"parentChildren": [{
								"identifier": "Location/109211",
								"childIdentifiers": ["Location/136256"]
							}, {
								"identifier": "Location/136256",
								"childIdentifiers": ["Location/137151"]
							}]
						}
					},
					"locationId": "136256"
				}],
				"practitionerId": ["136252"]
			}
		}
	}]
}

**Comment by dubdabasoduba** [comment]
> @rehammuzzamil Kindly share the official payload for this CC @dubdabasoduba, Also should @FikriMilano build on this PR #890?

@FikriMilano Yes, you can build on [PR 890](https://github.com/opensrp/fhircore/pull/890) however, I would suggest cherry-picking on the PR and then rebuilding it on the new code base. That's if we notice the conflicts fixes will be a bit complicated.

**Comment by dubdabasoduba** [comment]
> @RaaziaTarique Please use this Parameter resources sample to hold the Practitioner resource linkage.
> 
> ```
> {
>   "resourceType": "Parameters",
>   "parameter": [
>     {
>       "name": "practitioner",
>       "valueString": "003a78c1-a589-435d-8d28-12b3a77b471c", --- Practitioner identifier
>       "part": [
>         {
>           "name": "Organization",
>           "valueString": "214" 
>         },
>         {
>           "name": "CareTeam",
>           "valueString": "21"
>         },
>         {
>           "name": "CareTeam",
>           "valueString": "213"
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> * In the `part` attribute we have an array of objects. The attributes in each object are to  be used as follows
>   
>   * `name` -- this will hold the resource type
>   * `valueString` -- This will hold the resource ID.
> * After the generation of this Parameter resource we will save it and get its ID.
> * We will then save this ID in the secure shared preferences under the tag `practitioner_details`

@FikriMilano I think we might also want to change this implementation. I think we can just store the ids directly to the shared preferences. That helps us avoid too many searches. 

So store the ids using a `key` and `value` pair with the value being an array to cater for the user being assigned multiple times. Here is an example `practitioner-organisation : ["123", "345"]`

Let me know if this makes sense

**Comment by FikriMilano** [comment]
Yeah, that will work as well.
But for convention, I suggest we use camel case as the key naming format. e.g `practitioner_organization`

**Comment by FikriMilano** [comment]
A question.
So, we extract those organization, location, and care team IDs from the payload?
Then save it to the secure shared preference?

**Comment by FikriMilano** [comment]
Correct me if I'm wrong or missing something.
My thoughts process would be:

1. User logins
2. Returns payload
3. Extract IDs and keys to the secure shared preference
4. Extract attached FHIR resource and save it to the DB

**Comment by dubdabasoduba** [comment]
> A question. So, we extract those organization, location, and care team IDs from the payload? Then save it to the secure shared preference?

@FikriMilano Yes, we extract the resources and save them on the resource table. then extract the ids and save them in the shared preferences.

**Comment by dubdabasoduba** [comment]
> Correct me if I'm wrong or missing something. My thoughts process would be:
> 
> 1. User logins
> 2. Returns payload
> 3. Extract attached FHIR resources and save it to the DB
> 4. Extract resource IDs and save it to the secure shared preference

Yes, this is correct.

**Comment by dubdabasoduba** [comment]
@FikriMilano We can also add another mapping to the shared preference. 
 - The payload from the server has a custom resource called LocationHierarchy. Could we also save this in that shared pretences the key can be `practitioner_hierarchy`

**Comment by FikriMilano** [comment]
Wait, would it be better to have the location hierarchy as Parameters resource?
We can simply get the location recursively.

I think it's possible.
```
{
  "resourceType": "Parameters",
  "id": "xyz",
  "parameter": [
    {
      "name": "Location Hierarchy",
      "part": [
        {
          "name": "Charlie Clinic",
          "valueReference": "Location/109211",
          "part": [
            {
              "name": "Nala Location",
              "valueReference": "Location/136256",
              "part": [
                {
                  "name": "Test Village",
                  "valueReference": "Location/137151"
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by FikriMilano** [comment]
TODOs:

- [ ] Refactor practitioner details implementation to follow:
  - [ ] The latest payload
  - [ ] To store the IDs directly into the secure shared preference
  - [ ] Latest codebase
- [ ] R&D on Location Hierarchy using Parameters resource

cc @dubdabasoduba

**Comment by FikriMilano** [comment]
@dubdabasoduba 
I notice there is 2 API call response that has the same response.

- The first is from `protocol/openid-connect/userinfo` that returns this response:

```
{
  "sub": "b87ff3c2-cbc6-43e6-b753-a9620756f9e4",
  "email_verified": false,
  "realm_access": {
    "roles": [
      "realm-admin",
      "OPENMRS",
      "EDIT_KEYCLOAK_USERS",
      "VIEW_KEYCLOAK_USERS",
      "PLANS_FOR_USER",
      "offline_access",
      "uma_authorization",
      "ALL_EVENTS"
    ]
  },
  "organization": "105",
  "name": "Demo 16909",
  "groups": [
    "realm-admin",
    "OPENMRS",
    "EDIT_KEYCLOAK_USERS",
    "VIEW_KEYCLOAK_USERS",
    "PLANS_FOR_USER",
    "offline_access",
    "uma_authorization",
    "ALL_EVENTS"
  ],
  "preferred_username": "demo",
  "given_name": "Demo",
  "family_name": "16909"
}
```

- The second is from https://fhir.labs.smartregister.org/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9 which returns the same data as well:

```
{
  "KeycloakUserDetails": {
    "id": "de9ad01f-c485-4da4-9a81-284518e103c1",
    "user-bio": [
      {
        "identifier": "de9ad01f-c485-4da4-9a81-284518e103c1",
        "userName": "practitionera",
        "preferredName": "practitionera",
        "familyName": "Tester",
        "givenName": "Practitioner A",
        "emailVerified": "false"
      }
    ],
    "user-roles": [
      "ROLE_offline_access",
      "ROLE_default-roles-fhir-core",
      "ROLE_uma_authorization"
    ]
  }
}
```

At the current implementation https://github.com/opensrp/fhircore/pull/890, it's actually saving both, with different shared preference key.
This results in duplication and might potentially inconsistency between data.

So my question is.
Which one are we supposed to save into the shared preference?

**Comment by dubdabasoduba** [comment]
@FikriMilano I think the first one was added as a stopgap since we didn't have the practitioner details. 

I would say let's deprecate the first one. Let's also deprecate that API call. i.e we can have the following flow
1. Login
2. Hit the practitioner details then saving of that data takes over.

**Comment by FikriMilano** [comment]
Yes that would be ideal.

But from what I see, the first API call is the only way to get the keycloakUuid which is needed to call the practitioner details API.

**Comment by FikriMilano** [comment]
@dubdabasoduba 
I think my question would be.
Where does a keycloakUuid supposed to come from? or generated

**Comment by dubdabasoduba** [comment]
@FikriMilano good catch. is the payload returned after login? If not does login return anything else when successful?

**Comment by FikriMilano** [comment]
@dubdabasoduba 
No, the payload is not returned after login.

After successful login it returns only these folks: 
1. Token Name
2. Access Token
3. Token Type
4. Bearer
5. expires_in
6. refresh_expires_in
7. refresh_token
8. id_token
9. not-before-policy
10. session_state
11. scope

So, no keycloakUuid.

**Comment by FikriMilano** [comment]
I think the first API call only supposed to return a working dummy keycloakUuid.
I believe it's time to have the real thing, given that each Practitioner supposed to have their own keycloak.

Perhaps something like:
After successful login, also returns matching Practitioner username keycloak.

We are using this endpoint for login `protocol/openid-connect/token`

**Comment by dubdabasoduba** [comment]
> @dubdabasoduba @f-odhiambo It's a different issue compared to above, but this ticket is blocked by [opensrp/fhir-common-utils#5](https://github.com/opensrp/fhir-common-utils/issues/5)

@FikriMilano Do you mind explaining why the custom extension is needed? The Payload only has 2 custom data models i.e keycloak details and the LocationHeirarchy I think I am a bit lost by the Organization and CareTeam ones.

**Comment by dubdabasoduba** [comment]
> I think the first API call only supposed to return a working dummy keycloakUuid. I believe it's time to have the real thing, given that each Practitioner supposed to have their own keycloak.
> 
> Perhaps something like: After successful login, also returns matching Practitioner username keycloak.
> 
> We are using this endpoint for login `protocol/openid-connect/token`

This makes sense. It just seems like a duplication of effort to call both and return the keycloak details.

**Comment by FikriMilano** [comment]
> @FikriMilano Do you mind explaining why the custom extension is needed? The Payload only has 2 custom data models i.e keycloak details and the LocationHeirarchy I think I am a bit lost by the Organization and CareTeam ones. 


Yeah, we need those custom extension for parsing purpose.

Basically, any custom resource needs to be registered to the fhirParser.

And I wasn't able to access those 2 classes.

**Comment by FikriMilano** [comment]
> This makes sense. It just seems like a duplication of effort to call both and return the keycloak details. 

No worries.
Merging those 2 calls is actually better.

**Comment by FikriMilano** [comment]
Cool, do you have an ETA for the updated API call? the login fetchToken that includes payload

**Comment by FikriMilano** [comment]
@dubdabasoduba 
Could you create a ticket to track the merging of `protocol/openid-connect/token` and `practitioner-details?keycloak-uuid=` API call?

**Comment by dubdabasoduba** [comment]
@FikriMilano I think we can have both the calls for now. I will be creating another issue to enhance this. In that issue we will remove the keycloak response from the practitioner details.   

This `protocol/openid-connect/token` is a keycloak URL so merging it might be complicated.

**Comment by FikriMilano** [comment]
@dubdabasoduba 
Thanks, that's ok.

I would request for https://keycloak-stage.smartregister.org/auth/realms/FHIR_Android/protocol/openid-connect/userinfo to return this keycloak `34f0d616-afc8-4446-ae87-bb60be4bdbc9`
Since that's the keycloak we use for this payload
https://fhir.labs.smartregister.org/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9

Also, in the payload I request for the `Organization` that we use is `105` because the current one is still empty/no resource to be shown.

**Comment by FikriMilano** [comment]
I don't think it's possible to update it on my end. So I need your help with those.

**Comment by dubdabasoduba** [comment]
> @dubdabasoduba Thanks, that's ok.
> 
> I would request for https://keycloak-stage.smartregister.org/auth/realms/FHIR_Android/protocol/openid-connect/userinfo to return this keycloak `34f0d616-afc8-4446-ae87-bb60be4bdbc9` Since that's the keycloak we use for this payload https://fhir.labs.smartregister.org/fhir/practitioner-details?keycloak-uuid=34f0d616-afc8-4446-ae87-bb60be4bdbc9
> 
> Also, in the payload I request for the `Organization` that we use is `105` because the current one is still empty/no resource to be shown.

@FikriMilano I don't think you need to request for the `Organization`. That should be part of the FHIR practitioner details endpoint.  Which user are you using?

**Comment by FikriMilano** [comment]
@dubdabasoduba 
This one `34f0d616-afc8-4446-ae87-bb60be4bdbc9`.

What I mean is, this current Organization/`136254` has no proper resource to be shown, like household etc.

That's why I request it to be updated to `105`, which has that.

<img width="655" alt="Screen Shot 2022-08-15 at 20 57 14" src="https://user-images.githubusercontent.com/62053304/184649175-90ecfead-10b3-4cf4-bb25-74bf76b6cff5.png">

**Comment by FikriMilano** [comment]
I mean, if that's intentional, then it's ok.

**Comment by dubdabasoduba** [comment]
@FikriMilano That's okay we can change the assignment from FHiR web

**Comment by dubdabasoduba** [comment]
This issue is just handling the saving and reading of the practitioner-related data. After these ones are complete then the fetching of data based on what you save here will be completed

**Comment by FikriMilano** [comment]
Yeah that's right, thanks

**Comment by FikriMilano** [comment]
TODOs:

- [x] Refactor practitioner details implementation to follow: #1520 
  - [x] Latest codebase
  - [x] Fix broken implementation, after code breaking changes from latest codebase
  - [x] The latest payload
  - [x] To store the IDs directly into the shared preference
 
- [ ] Fix failing test and add more test https://github.com/opensrp/fhircore/pull/890

cc @f-odhiambo @dubdabasoduba

**Comment by abbeyseto** [comment]
> > We tried the possibility to add a custom UserDetails Resource class
> > ```
> > import org.hl7.fhir.r4.model.Resource;
> > import org.hl7.fhir.r4.model.ResourceType;
> > 
> > public class UserDetails extends Resource {
> >   @Override
> >   public ResourceType getResourceType() {
> >     return ResourceType.Basic;
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But you see we need to extend getResourceType() method and it can only use ResourceType which is only available in the Fhir library. Even if we want make modifications into extending resource type we need to use pre-defined ResourceType.
> > <img alt="Screenshot 2021-12-24 at 1 09 42 PM" width="1092" src="https://user-images.githubusercontent.com/89384079/147351642-cbe8246e-2ca3-4f79-9268-76c06bcefc04.png">
> > @dubdabasoduba @rehammuzzamil @maimoonak
> 
> @RaaziaTarique As discussed, we need to register our custom Resource Provider in the FHIR. On the server-side, we achieve this by creating a new class with the `@ResourceDef `annotation. Once the class is created, we need to register our new Resource Provider (custom class) and other supporting classes in the Restful server. Registration of Resource Provider can be done by using `registerProvider(practitionerDetailsResourceProvider); `method at the backend.
> 
> Following is the example of custom resource provider:
> 
> ```
> 
> @ResourceDef(
>         name = "practitioner-details",
>         profile = "http://hl7.org/fhir/profiles/custom-resource")
> public class PractitionerDetails extends Practitioner {
> 
>     @Child(
>             name = "KeycloakUserDetails",
>             type = {KeycloakUserDetails.class})
>     @Description(
>             shortDefinition = "Complete KeycloakUserDetails",
>             formalDefinition = "KeycloakUserDetails")
>     private KeycloakUserDetails keycloakUserDetails;
> 
>     @Child(
>             name = "fhir",
>             type = {FhirPractitionerDetails.class})
>     @Description(
>             shortDefinition = "Get data from FHIR Server",
>             formalDefinition = "Get data from FHIR Server")
>     private FhirPractitionerDetails fhirPractitionerDetails;
> 
>     @Override
>     public Practitioner copy() {
>         Practitioner practitioner = new Practitioner();
>         Bundle bundle = new Bundle();
>         List<Bundle.BundleEntryComponent> theEntry = new ArrayList<>();
>         Bundle.BundleEntryComponent entryComponent = new Bundle.BundleEntryComponent();
>         entryComponent.setResource(new Bundle());
>         theEntry.add(entryComponent);
>         bundle.setEntry(theEntry);
>         this.copyValues(practitioner);
>         return practitioner;
>     }
> 
>     @Override
>     public ResourceType getResourceType() {
>         return ResourceType.Bundle;
>     }
>   //getters and setters for the child fields
> ```
> 
> }
> 
> Since the work is in progress. It is likely to be changed. cc: @dubdabasoduba @maimoonak

I am using hapi-fhir-jpaserver-starter, and I'm a bit confused about where these resources will be created in the folder structure. I don't seem to see resource definition files. @dubdabasoduba @FikriMilano
